### DISCUSSION TITLE
`vercel pull` pulls all preview environment variables, including custom branches

### DISCUSSION THREAD

**Discussion by kev-cao** [vercel/vercel#8652] [Help] [answered]
When I do `vercel pull --environment=preview` with the CLI, it pulls **all** environment variables under the `preview` environment, including ones set for specific GitHub branches. Is there a way to pull only the standard set of `preview` environment variables without pulling the ones specific to certain branches?

**Comment by amyegan** [comment]
Hi @defCoding! I don't know of a way to specify a branch or standard set of `preview` environment variables, so reached out to the CLI team to see if they have a solution for this. Will update here when I know more.

**Reply by amyegan** [reply]
Confirmed with the team that we do not currently have a way to specify a branch or standard set of `preview` environment variables. No ETA yet, but they are exploring options to add this ability. Sorry there is no fix ready at the moment 😞

**Reply by kev-cao** [reply]
Ah that's fine! I found a bit of a fix around it by just creating a script to modify the `.vercel/.env.preview.local` after a `vercel pull`. Just wanted to make sure I wasn't missing something. Thanks for the clarification!

**Reply by giancarlosisasi** [reply]
hey @defCoding could u please share what script are u using to edit the .local/.prod file? 🙏🏼

**Reply by giancarlosisasi** [reply]
Also, @amyegan I think the problem could be considered a bug since `vercel pull` is downloading and overriding the env var in different preview branches 😢 . 
For example, I have two env vars called CND, one for preview and one for the QA branch. When using `vercel pull --environment=preview` it's downloading the value for the QA branch and not for the preview env itself 😭

**Reply by kev-cao** [reply]
> hey @defCoding could u please share what script are u using to edit the .local/.prod file? 🙏🏼

It's nothing particularly fancy haha, there was an extra variable that was being included that shouldn't have been, so all it does is this:

```bash
grep -Ev "^VARIABLE=" .vercel/.env.preview.local > temp && mv temp .vercel/.env.preview.local
```

If there is a list of unwanted variables, it should be trivial to create a file `.envignore` or something that a script can read from and run a `grep -Ev "^(VARIABLE1|VARIABLE2|...)"` on `.vercel/.env.preview.local`.

**Reply by giancarlosisasi** [reply]
alright, thanks a lot @defCoding !

**Comment by amyegan** [comment, marked as answer]
Hi everyone, I have some good news to share! A `--git-branch` option is available on `vc pull` and `vc env pull` in [vercel@28.9.0](https://github.com/vercel/vercel/releases/tag/vercel%4028.9.0).

So now you can use `vercel pull --environment=preview --git-branch={branch-name}` to pull preview environment variables for a specific branch! 🥳

**Reply by Barefoot-Dev** [reply]
If I have an env var set for a all preview branches, but then I customize it for once specific branch and try to vc pull that branch, it seems to pull the env variable set to preview, rather than the custom branch. Is this expected?

**Comment by mazharz** [comment]
@amyegan 

I noticed that the `--git-branch` didn't appear on the docs nor in the output of `vercel help pull` or equivalently `vercel pull --help`. I would like to add that to this file which I believe is used for showing the CLI options: https://github.com/vercel/vercel/blob/main/packages/cli/src/commands/pull/command.ts#L5

I would also like to add it to the docs on the website, but I am not able to find the docs repo, please let me know if this change is ok with the maintainers and if I am able to add it to the docs.

**Reply by amyegan** [reply]
Thanks for sharing this! The [CLI docs](https://vercel.com/docs/cli) are not publicly editable, but I shared your feedback with the team. If you have any other suggestions, please let me know and I'll make sure it reaches the right people. 🙏
